crates/fluvio/src/consumer/retry.rs

```diff
@@ -90,70 +94,79 @@ impl ConsumerWithRetry {
         self.state = new_state;
         self.notify.notify_one();
     }
+
+    fn set_idle(&mut self) {
+        self.change_state(ConsumerState::Idle);
+        self.notify.notify_one();
+    }
+
+    fn set_terminated(&mut self) {
+        self.change_state(ConsumerState::Terminated);
+        self.notify.notify_one();
+    }
+
+    fn set_task(&mut self, task: BoxConsumerFuture) {
+        self.change_state(ConsumerState::Task(task));
+    }
 }
 
 impl Stream for ConsumerWithRetry {
     type Item = Result<ConsumerRecord, ErrorCode>;
 
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        // We use a loop so that after a state transition we immediately poll again.
-        let this = self.get_mut();
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
```

Comment:
Can you check if it is possible to implement using `https://crates.io/crates/async-stream` instead of implementing custom polling?

Reply:
I think it's possible, but I need to rewrite all implementation of `ConsumerStream` it needs Stream from `futures-core`

Reply:
I propose a new PR for this technical debt

Reply:
here: https://github.com/infinyon/fluvio/issues/4406